README
==== NAME ====

LCFG::Build::Tools - Tools for managing the release of LCFG projects

==== DESCRIPTION ====

LCFG::Build::Tools is a suite of tools designed to handle the
releasing of LCFG software projects and the creation of
packages. Support is available for developing projects within a
version-control systems (currently either CVS or None). By default a
source tar file is generated along with a specfile for building binary
RPMs. Support is provided for building binary RPMs directly. Work is
under way to also fully support the generation of MacOSX packages.

Although this software is designed for managing LCFG projects there is
nothing that requires the software be for LCFG. All the tools included
are designed to more widely applicable.

This suite has been intentionally designed to be easy to extend to
support new version-control systems (e.g. subversion, git, etc) and
new package formats (e.g. for Debian). It has also been designed to
ensure that it is easy to extend with additional command modules. For
further details see the online documentation at
http://www.lcfg.org/doc/buildtools/

==== DEPENDENCIES ====

This module needs perl 5.8.1 or newer (that's for Moose nothing in
this code needs it).

To build this module you will need the following:

        Module::Build

To use this module you will need the following:

        Archive::Tar
        DateTime
        File::Find::Rule
        IO::Zlib
        LCFG::Build::PkgSpec >= 0.0.22
        LCFG::Build::VCS >= 0.0.19
        Moose >= 0.57
        MooseX::App::Cmd >= 0.06
        Template Toolkit >= 2.14
        UNIVERSAL::require
        YAML::Syck >= 0.98

==== INSTALLATION ====

Make sure you have the dependencies installed first! (see DEPENDENCIES above)

If you don't know how to install these, try using the CPAN module, an easy
way of auto-installing modules from the Comprehensive Perl Archive Network,
where the above modules reside. Do "perldoc perlmodinstall" or "perldoc
CPAN" for more information.

To install this module type the following:

   perl Build.PL
   ./Build
   ./Build test
   ./Build install

Or, if you're on a platform (like DOS or Windows) that doesn't like the "./"
notation, you can do this:

   perl Build.PL
   perl Build
   perl Build test
   perl Build install

In order to install somewhere other than the default, such as in a directory
under your home directory, like "/home/fred/perl" go

   perl Build.PL --install_base /home/fred/perl

as the first step instead.

This will install the files underneath /home/fred/perl.

You will also need to ensure the lcfgbuild files are installed into a
writable location. So you will normally need something like:

   ./Build --install_base /home/fred/perl \
           --install_path lcfgbuild=/home/fred/lcfgbuild

If they are not in the standard location the LCFG::Build::Utils
module needs to know where the templates are stored. You can override
the default location with the LCFG_BUILD_TMPLDIR environment variable,
for example:

        export LCFG_BUILD_TMPLDIR=/home/fred/lcfgbuild/templates

You will also need to make sure that you alter the PERL5LIB variable
to find the modules, and the PATH variable to find the script.

Therefore you will need to change: your path, to include
/home/fred/perl/script (where the script will be)

        export PATH=/home/fred/perl/script:${PATH}

the PERL5LIB variable to add /home/fred/perl/lib

        export PERL5LIB=/home/fred/perl/lib:${PERL5LIB}

If you want to install in a temporary install directory (such as if you are
building a package) then instead of going

   perl Build install

go

   perl Build install destdir=/my/temp/dir

and it will be installed there, with a directory structure under
/my/temp/dir the same as it would be if it were installed plain. Note that
this is NOT the same as setting --install_base, because certain things are
done at build-time which use the install_base info.

See "perldoc perlrun" for more information on PERL5LIB, and see "perldoc
Module::Build" for more information on installation options.

==== PROBLEMS AND FEEDBACK ====

There are no known bugs in this application. Please report any
problems to bugs@lcfg.org, feedback and patches are also always very
welcome.

==== AUTHOR ====

Stephen Quinney <squinney@inf.ed.ac.uk>

==== COPYRIGHT AND LICENCE ====

Copyright (C) 2008 University of Edinburgh

This program is free software; you can redistribute it and/or modify it
under the terms of the GPL, version 2 or later.



Changes
2014-10-31  SVN: new release

	* Release: 0.4.5

2014-10-31 10:52  squinney@INF.ED.AC.UK

	* lib/LCFG/Build/Utils/OSXPkg.pm.in: Fixed exit with a warning
	  about the tmp dir, bug#791

2014-07-02 07:38  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.4.4

2014-07-02 07:38  squinney@INF.ED.AC.UK

	* templates/lcfg.cmake.tt: Added CentOS support

2014-07-01 16:05  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.4.3

2014-07-01 16:05  squinney@INF.ED.AC.UK

	* LCFG-Build-Tools.spec: Added missing build dep

2014-07-01 16:01  squinney@INF.ED.AC.UK

	* templates/lcfg.cmake.tt: Added support for detecting RHEL release

2014-04-01 14:43  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.4.2

2014-04-01 14:43  squinney@INF.ED.AC.UK

	* templates/build.cmake.tt: fixed missing closing brackets

2014-04-01 14:42  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.4.1

2014-04-01 14:42  squinney@INF.ED.AC.UK

	* templates/build.cmake.tt: only attempt to install a man page
	  created using pod2man if it actually generated a file

2013-10-11 12:54  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.4.0

2013-10-11 12:53  squinney@INF.ED.AC.UK

	* LCFG-Build-Tools.spec, META.yml.in, Makefile.PL, lcfg.yml,
	  lib/LCFG/Build/Tool.pm.in: Converted the detection of the
	  revision control system to use the new auto_detect methods
	  provided by the LCFG::Build::VCS modules in version 0.2.0. This
	  should fix auto-detection for svn 1.7 and newer, see
	  https://bugs.lcfg.org/show_bug.cgi?id=685

2013-05-14 08:12  squinney@INF.ED.AC.UK

	* lib/LCFG/Build/Tool/CheckMacros.pm.in: Added new macros to the
	  basic list

2013-05-14 08:09  squinney@INF.ED.AC.UK

	* lcfg_config.yml: Added BOOTSTAMP and RELEASEFILE. This is
	  primarily for building the LCFG client but other things might
	  benefit as well

2012-05-18 17:03  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.3.1

2012-05-18 17:03  squinney@INF.ED.AC.UK

	* Build.PL.in, LCFG-Build-Tools.spec, META.yml.in, Makefile.PL,
	  README: Added the missing dependency on the Readonly module

2012-05-18 16:54  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.3.0

2012-05-18 16:54  squinney@INF.ED.AC.UK

	* ., LCFG-Build-Tools.spec, MANIFEST, META.yml.in, lcfg.yml,
	  lib/LCFG/Build/Tool/DevOSXPkg.pm.in,
	  lib/LCFG/Build/Tool/OSXPkg.pm.in,
	  lib/LCFG/Build/Utils/MacOSX.pm.in,
	  lib/LCFG/Build/Utils/OSXPkg.pm.in, t/01_load.t: Implemented
	  creating Apple Packages via the built in "pkgbuild" tool.
	  This initial commit can only process projects that use CMake to
	  build
	  and install. Thanks go to Kenny MacDonald for adding this new
	  functionality. This change closes
	  https://bugs.lcfg.org/show_bug.cgi?id=565

2011-12-22 15:26  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.2.2

2011-12-22 15:26  squinney@INF.ED.AC.UK

	* templates/lcfg.cmake.tt: Applied patch to fix MacOSX support, see
	  bug#304 for details

2011-11-09 12:47  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.2.1

2011-11-09 12:46  squinney@INF.ED.AC.UK

	* lcfg_config.yml, lib/LCFG/Build/Tool/CheckMacros.pm.in: Added
	  LCFGRUN which is a macro which refers to a directory where
	  component run files can be stored. Currently this is the same as
	  LCFGTMP for backwards-compatibility reasons

2011-03-03 20:36  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.2.0

2011-03-03 20:30  squinney@INF.ED.AC.UK

	* lib/LCFG/Build/Utils/RPM.pm.in: The parsing and formatting of the
	  changelogs has been completely rewritten to make it more robust

2011-03-02 12:20  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.1.6

2011-03-02 12:20  squinney@INF.ED.AC.UK

	* lcfg.yml, templates/lcfg.cmake.tt: Fixed SL6 support, rather
	  annoyingly they have decided to change their LSB distributor ID
	  string

2011-02-18 13:30  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.1.5

2011-02-18 13:30  squinney@INF.ED.AC.UK

	* lib/LCFG/Build/Tool/Submit.pm.in: Fixed submit tool typos

2011-02-17 18:15  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml, lib/LCFG/Build/Tool/Submit.pm.in:
	  LCFG-Build-Tools release: 0.1.4

2011-02-17 18:13  squinney@INF.ED.AC.UK

	* lib/LCFG/Build/Tool.pm.in: Fixed typo

2011-02-17 18:11  squinney@INF.ED.AC.UK

	* templates/lcfg.cmake.tt: need to escape double-quotes in cmake
	  strings

2011-02-17 17:01  squinney@INF.ED.AC.UK

	* lib/LCFG/Build/Tool/DevRPM.pm.in, lib/LCFG/Build/Tool/RPM.pm.in:
	  Added docs about the new --sign option

2011-02-16 14:40  squinney@INF.ED.AC.UK

	* lib/LCFG/Build/Tool.pm.in, lib/LCFG/Build/Tool/DevRPM.pm.in,
	  lib/LCFG/Build/Tool/RPM.pm.in, lib/LCFG/Build/Utils/RPM.pm.in:
	  Added ability to pass the --sign option through to rpmbuild from
	  the devrpm and rpm commands. This resolves
	  https://bugs.lcfg.org/show_bug.cgi?id=387

2010-10-11 12:51  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.1.3

2010-10-11 12:45  squinney@INF.ED.AC.UK

	* lcfg.yml, lib/LCFG/Build/Tool/DevRPM.pm.in: Set the build
	  directory so that we can keep the build directory when making
	  devel packages. This makes it easier to examine what has been
	  generated

2010-10-11 12:44  squinney@INF.ED.AC.UK

	* lib/LCFG/Build/Utils/RPM.pm.in: Allow the calling code to specify
	  the rpm BUILD directory

2010-07-20 13:13  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.1.2

2010-07-20 13:13  squinney@INF.ED.AC.UK

	* lcfg.yml, lib/LCFG/Build/Tool/DevPack.pm.in,
	  lib/LCFG/Build/Tool/Pack.pm.in, lib/LCFG/Build/Utils.pm.in:
	  Report the full path to the generated tar file when using pack
	  and devpack. This fixes https://bugs.lcfg.org/show_bug.cgi?id=293

2010-07-20 13:08  squinney@INF.ED.AC.UK

	* Changes, lcfg.yml: LCFG-Build-Tools release: 0.1.1


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.03
use ExtUtils::MakeMaker;
WriteMakefile
(
          'PL_FILES' => {},
          'INSTALLDIRS' => 'site',
          'NAME' => 'LCFG::Build::Tools',
          'EXE_FILES' => [
                           'bin/lcfg-reltool'
                         ],
          'VERSION_FROM' => 'lib/LCFG/Build/Tools.pm',
          'PREREQ_PM' => {
                           'LCFG::Build::VCS' => '0.0.19',
                           'MooseX::App::Cmd' => '0.06',
                           'File::HomeDir' => '0.58',
                           'Archive::Tar' => '0',
                           'Moose' => '0.57',
                           'Module::Pluggable' => '3.10',
                           'File::Find::Rule' => '0',
                           'Test::Exception' => 0,
                           'IO::Zlib' => '0',
                           'YAML::Syck' => '0.98',
                           'Template' => '2.14',
                           'DateTime' => 0,
                           'Readonly' => '0',
                           'Test::More' => 0,
                           'Test::Differences' => 0,
                           'LCFG::Build::PkgSpec' => '0.0.22',
                           'UNIVERSAL::require' => '0'
                         }
        )
;


Build.PL
#!/usr/bin/perl
use Module::Build; # -*-cperl-*-

my %yml_files = map { $_ => "lcfgbuild/$_" }
    qw(lcfg_config.yml mapping_config.yml);

my %tmpl_files = map { $_ => "lcfgbuild/$_" }
    ( glob('templates/*.tt'), glob('templates/*.tmpl') );

my $build = Module::Build->new(
    module_name   => 'LCFG::Build::Tools',
    dist_abstract => 'LCFG build system tools',
    dist_author   => 'Stephen Quinney <squinney@inf.ed.ac.uk>',
    dist_version  => '0.4.5',
    license => 'gpl',
    requires => {
        'perl'                 => '5.6.1',
        'Archive::Tar'         => '0',
        'DateTime'             => '0',
        'File::HomeDir'        => '0.58',
        'File::Find::Rule'     => '0',
        'IO::Zlib'             => '0',
        'LCFG::Build::PkgSpec' => '0.0.22',
        'LCFG::Build::VCS'     => '0.0.19',
        'Module::Pluggable'    => '3.10',
        'Moose'                => '0.57',
        'MooseX::App::Cmd'     => '0.06',
        'Template'             => '2.14',
        'Readonly'             => '0',
        'UNIVERSAL::require'   => '0',
        'YAML::Syck'           => '0.98',
    },
    build_requires => {
        'Test::More'        => 0,
        'Test::Differences' => 0,
        'Test::Exception'   => 0,
    },
    add_to_cleanup => [ 'LCFG-Build-Tools-*', 'MANIFEST.bak', 'Changes.bak' ],
    create_makefile_pl => 'traditional',
    yml_files    => \%yml_files,
    tmpl_files   => \%tmpl_files,
    install_path => { lcfgbuild => '/usr/share/lcfgbuild' },
);
$build->add_build_element('yml');
$build->add_build_element('tmpl');
$build->create_build_script;
